libs/core/langchain_core/tools.py

```diff
@@ -511,67 +512,37 @@ def run(
-            parsed_input = self._parse_input(tool_input)
-            tool_args, tool_kwargs = self._to_args_and_kwargs(parsed_input)
-            observation = (
-                context.run(
-                    self._run, *tool_args, run_manager=run_manager, **tool_kwargs
-                )
-                if new_arg_supported
-                else context.run(self._run, *tool_args, **tool_kwargs)
+            tool_args, tool_kwargs = self._to_args_and_kwargs(tool_input)
+            content, raw_output = context.run(
+                self._run_include_raw_output,
+                *tool_args,
+                run_manager=run_manager,
+                **tool_kwargs,
             )
         except ValidationError as e:
-            if not self.handle_validation_error:
-                raise e
-            elif isinstance(self.handle_validation_error, bool):
-                observation = "Tool input validation error"
-            elif isinstance(self.handle_validation_error, str):
-                observation = self.handle_validation_error
-            elif callable(self.handle_validation_error):
-                observation = self.handle_validation_error(e)
-            else:
-                raise ValueError(
-                    f"Got unexpected type of `handle_validation_error`. Expected bool, "
-                    f"str or callable. Received: {self.handle_validation_error}"
-                )
-            return observation
+            content = _handle_validation_error(e, flag=self.handle_validation_error)
         except ToolException as e:
             if not self.handle_tool_error:
                 run_manager.on_tool_error(e)
                 raise e
-            elif isinstance(self.handle_tool_error, bool):
-                if e.args:
-                    observation = e.args[0]
-                else:
-                    observation = "Tool execution error"
-            elif isinstance(self.handle_tool_error, str):
-                observation = self.handle_tool_error
-            elif callable(self.handle_tool_error):
-                observation = self.handle_tool_error(e)
-            else:
-                raise ValueError(
-                    f"Got unexpected type of `handle_tool_error`. Expected bool, str "
-                    f"or callable. Received: {self.handle_tool_error}"
-                )
-            run_manager.on_tool_end(observation, color="red", name=self.name, **kwargs)
-            return observation
+            content = _handle_tool_error(e, flag=self.handle_tool_error)
+            run_manager.on_tool_end(content, color="red", name=self.name, **kwargs)
         except (Exception, KeyboardInterrupt) as e:
             run_manager.on_tool_error(e)
             raise e
         else:
-            run_manager.on_tool_end(observation, color=color, name=self.name, **kwargs)
-            return observation
+            run_manager.on_tool_end(content, color=color, name=self.name, **kwargs)
```

Comment:
This affects the astream events API. I'd argue the raw output is more important in that case than the stringified version of it.

The least surprising thing is either ToolMessage or else the 2-tuple output?

Reply:
could log whatever actual output is (so ToolMessage if thats whats returned below)

Reply:
Let's do `ToolMessage`